Snowflake.Data/Core/Session/SessionPool.cs

```diff
@@ -95,68 +100,154 @@ internal SFSession GetSession(string connStr, SecureString password)
+            if (currentSize < _maxPoolSize)
+            {
+                s_logger.Debug($"SessionPool - new session creation granted because current size is {currentSize} out of {_maxPoolSize}");
+                return true;
+            }
+            s_logger.Debug($"SessionPool - could not grant new session creation because current size is {currentSize} out of {_maxPoolSize}");
+            return false;
+        }
+        
+        private SFSession WaitForSession(string connStr)
+        {
+            var timeout = _waitingQueue.GetWaitingTimeoutMillis();
+            s_logger.Debug($"SessionPool::WaitForSession for {timeout} millis timeout");
+            var beforeWaitingTime = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
+            long nowTime = beforeWaitingTime;
+            while (nowTime < beforeWaitingTime + timeout) // we loop to handle the case if someone overtook us after being woken or session which we were promised has just expired
+            {
+                var timeoutLeft = beforeWaitingTime + timeout - nowTime;
+                var successful = _waitingQueue.Wait((int) timeoutLeft, CancellationToken.None);
+                if (!successful)
+                {
+                    s_logger.Debug($"SessionPool::WaitForSession - woken without a session granted");
+                    throw WaitingFailedException();
+                }
+                s_logger.Debug($"SessionPool::WaitForSession - woken with a session granted");
+                lock (_sessionPoolLock)
+                {
+                    var session = ExtractIdleSession(connStr);
+                    if (session != null)
+                    {
+                        s_logger.Debug($"SessionPool::WaitForSession - a session was extracted from idle sessions");
+                        return session;
+                    }
+                }
+                nowTime = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
+            }
+            s_logger.Debug($"SessionPool::WaitForSession - could not find any idle session available withing a given timeout");
+            throw WaitingFailedException();
+        }
+
+        private static Exception WaitingFailedException() => new Exception("Could not obtain a connection from the pool within a given timeout");
+
+        private SFSession ExtractIdleSession(string connStr)
+        {
+            for (int i = 0; i < _idleSessions.Count; i++)
+            {
+                if (_idleSessions[i].ConnectionString.Equals(connStr))
+                {
+                    SFSession session = _idleSessions[i];
+                    _idleSessions.RemoveAt(i); // we don't do _waitingQueue.OnResourceDecrease() here because it happens in GetIdleSession()
+                    long timeNow = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
+                    if (session.IsExpired(_timeout, timeNow))
+                    {
+                        session.close(); // TODO: change to session required
```

Comment:
I changed the TODO comment to be more precise. But it is for purpose, because we need to cherry-pick/merge changes from SNOW-984600 here. But maybe not in this PR.